typing/ctype.ml

```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+   * It first makes a separate copy of the type as a graph, omitting nodes
+     that have no free univars.
+   * In this first pass, [visited] is used as a mapping for previously visited
+     nodes, and must already contain all the free univars in [ty].
+   * The remaining (univar-closed) parts of the type are then instantiated
+     with [copy] using a common [copy_scope].
+   The reason to work in two passes lies in recursive types such as:
+     [let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m]
+   The type of [x#m] should be:
+     [ < n : 'c; p : < m : 'a. < n : 'a; p : 'b > > as 'b > ]
+   I.e., the universal type variable ['a] is both instantiated as a fresh
+   type variable ['c] when outside of its binder, and kept as universal
+   when under its binder.
+   Assumption: in the first call to [copy_sep], all the free univars should
+   be bound by the same [Tpoly] node. This guarantees that they are only
+   bound when under this [Tpoly] node, which has no free univars, and as
+   such is not part of the separate copy. In turn, this allows the separate
+   copy to keep the sharing of the original type without breaking its
+   binding structure.
+ *)
+let rec copy_sep ~copy_scope ~fixed ~free ~may_share
+    ~(visited : type_expr TypeHash.t) (ty : type_expr) =
   let univars = free ty in
   if is_Tvar ty || may_share && TypeSet.is_empty univars then
     if get_level ty <> generic_level then ty else
     let t = newstub ~scope:(get_scope ty) in
-    delayed_copy :=
-      lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty)))
-      :: !delayed_copy;
+    add_delayed_copy t copy_scope ty;
     t
   else try
-    let t, bound_t = List.assq (get_id ty) visited in
-    let dl = if is_Tunivar ty then [] else diff_list bound bound_t in
-    if dl <> [] && conflicts univars dl then raise Not_found;
-    t
+    TypeHash.find visited ty
   with Not_found -> begin
     let t = newstub ~scope:(get_scope ty) in
-    let desc = get_desc ty in
-    let visited =
-      match desc with
-        Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
-          (get_id ty, (t, bound)) :: visited
-      | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
-          visited
-      | Tlink _ | Tsubst _ ->
-          assert false
-    in
-    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~bound visited in
+    TypeHash.add visited ty t;
+    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~visited in
     let desc' =
-      match desc with
+      match get_desc ty with
       | Tvariant row ->
           let more = row_more row in
           (* We shall really check the level on the row variable *)
           let keep = is_Tvar more && get_level more <> generic_level in
+          (* In that case we should keep the original, but we still
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
```

Comment:
`keep` being true tells us that the polymorphic variant has a non-generic row variable. If I understand the `add_delayed_copy` here, the idea is to `copy` the whole polyrmorphic variant in this case. Why is this correct? What if one of the variant fields contains one of the instantiated univars?

This is the key of the bugfix in this PR. I understand that you want to guarantee that if the Tvariant is not shared, then the row is also not shared. Here you do this by always sharing the Tvariant when you know that you need to share the row.

Instead I would expect the current `may_share` logic to be enough here. At this point we know that we will unshare the `Tvariant` node, so we want the row variable to be unshared as well. Why isn't `may_share:false` enough? Because the check for sharing nodes in `copy_sep` is `is_Tvar ty || may_share && is_empty univars`, so Tvar rows are shared even when `may_share;false`. Would removing the `is_Tvar ty` part of the condition fix the issue? (Obviously this would not make any difference when `may_share:true`, as Tvar nodes are univar-free.)

Reply:
Actually, this is not the main part of the fix. The main part is the switch from using `visited` only for recursion (and substitution of univars) to using it to keep sharing.
Now, for this specific case. If `row_more` is a non-generic type variable, then it means that structure of the row should be shared with the original. The safest way to do that is to keep the original one.
The previous code duplicated the row, but kept the sharing in the `row_field_desc`'s. This had the strange consequence that some function could be typable, but not applicable because as soon as you tried to unify it with something else as a univar would leak. I came to the conclusion that it was better to keep the original row than to try to create this kind of copy, in particular this avoid delaying errors.
Note that this mechanism of copying while keeping sharing with the original is also used in `Subst.typexp`. There the problem is a bit different because one no longer refers to the original after copying, so there `keep` acts more like a safety mechanism.

Reply:
I still wonder whether this is correct: could we have a variant with a non-generic row variable, but fields that refer to the instantiated univars? In this case, isn't the `keep=true` case iconrrect?

Reply:
This is indeed the case, and there are such examples in the test suite. The point is that we are going to fail immediately (when we check the scoping of univars on the result of the substitution), rather than later on as was the case before.

Reply:
Stated differently, a type whose row variable is a Tvar should not contain free univars to start with. Currently this is not enforced explicitly (this should probably be changed, but this concerns a different part of the code); we cannot fail inside `copy_sep` so we delay the failure until just after. Note also that if the row variable is a Tunivar, then all free univars should be bound outside of the scope of this Tunivar; this one is properly enforced.

Reply:
I don't understand, probably because my mental model of variants / row variables is wrong. Intuitively I see nothing wrong with having a non-generic row variable and univars in the field types. (We allow mixing non-generic *type* variables and univars, why should row variables be handled differently?) Does this assumption that they cannot happen come from the OCaml syntax for rows, where we cannot name the row variable itself and have to name the whole variant instead, capturing the field types as the same time?

That is: I think that

```
let f (x : [ | 'r]) (y : < m : 'b . [ `Foo of 'b | 'r ] >) = ignore y#m
```

should be accepted, but I guess we currently have to write this as

```
let f (x : 'a) (y : < m : 'b . ([> `Foo of 'b ] as 'a) >) = ignore y#m
```

and this is invalid / can never happen because `'b` escapes through `'a`. Is this the idea?

(But: aren't there ways to get a non-generic row type variable without naming/escaping the field types?)

Reply:
First, about why it is wrong to duplicate the fields of a row whose (normal) variable is not generic.
There is actually a mismatch between the implementation and the theory. The theory is based on Ohori's kinds, where a row variable is seen as representing the whole variant type, rather than just its extension as in Remy's approach, which is followed in the implementation. From this point of view, having a normal row variable capture a univar in one of the fields is incorrect in the theory. Even if we follow Remy's approach, there is a problem with the `Reither` fields. Those fields can be extended with extra types, that may eventually be unified together.  Sharing the row variable means also sharing the extension variable of all the `Reither` fields. If this field contains univars that are going to be instantiated, then it means that we end up someday unifying the extension with both the instantiated univars (which are normal vars) and the original ones, which will inevitably cause a clash, and a harder to understand one.
You can see this in the example `j` in the testsuite:
```ocaml
let j (x : < m : 'a. [< `A of 'a ] -> 'c > as 'c) = x#m;;
```
It used to be accepted, but you will have a hard time defining an object with this type, and even if you succeed, unification will fail when try to pass it to `j` or to use its result. The reason is that inside the type of `x#m` there are two instances `[< A of 'a]` sharing the same extension. This should be something like:
```ocaml
[< `A of 'b & 'e] -> (< m : 'a. [< `A of 'a & 'e] -> 'c > as 'c)
```
You see the contradiction between the two instances, as using the field `A` would result in unifying `'b` and `'a`, which is clearly impossible.
If we follow Remy's approach, it would be OK to allow univars in the `Rpresent` fields, but then we need to build the corresponding theory, and I think it is intrinsically more complex, for limited practical benefits. The reason we are still sticking with implementation is more related to the principality tracking: by using a different level for the `Tvariant` and `row_more` we can generalize the level of more nodes, i.e. see more types as known, which is useful in practice.

Now, about the `is_Tvar`. If the row variable is either `Tunivar` or `Tconstr`, it is OK to duplicate the row. Its contents are _fixed_ in the original, so that there is no need to share extension between the original and the copy. This is for instance what is happening in example `g`, where we simultaneously instantiate `'a` and the row variable `'b`:
```ocaml
let g (x : < m : 'a 'b. <n : ([< `A of 'a ] as 'b) -> 'c > > as 'c) = x#m;;
```

Reply:
Thanks for the explanation regarding the rows. I hadn't realized that the row variable in OCaml is supposed to represent the whole variant/object. (But then this morning I was tearing my hair trying to understand `Tobject` code which I guess makes more sense now.)

>  Even if we follow Remy's approach, there is a problem with the `Reither` fields. Those fields can be extended with extra types, that may eventually be unified together. Sharing the row variable means also sharing the extension variable of all the `Reither` fields.

I thought, when reading the code of `copy` and `copy_sep`, that doing this properly was the job of the `keep` parameter to `copy_row` -- that it would take into account, during field copy, the fact that the row variable is non-generic and will be shared, so that the field extensions are shared as well.

> Now, about the `is_Tvar`. If the row variable is either `Tunivar` or `Tconstr`, it is OK to duplicate the row. Its contents are fixed in the original, so that there is no need to share extension between the original and the copy.

I'm not sure you are replying to the question I was asking. I understand your answer as related to this occurrence of `is_Tvar`:

https://github.com/ocaml/ocaml/blob/f669914c672c8b77f24e96d81c6b7100039086ee/typing/ctype.ml#L1350

but I was meaning to ask about this one:

https://github.com/ocaml/ocaml/blob/f669914c672c8b77f24e96d81c6b7100039086ee/typing/ctype.ml#L1334

This occurrence of `is_Tvar` forces copying the row variable even when `may_share:false` is passed to the recursive calls of `copy_sep` in the Tvariant and Tfield case. I wonder why this is the correct thing to do. I may be missing something, but I have the impression that if we didn't have this `is_Tvar ty ||` condition we could safely use the same code in the Tvariant case whether `keep` is true or not (... assuming that `keep` is in charge of keeping/sharing the field extensions when the row variable is not generic), just like `copy` is doing. (And `copy` doesn't have an `is_Tvar` condition to decide whether to share or to duplicate.)

Reply:
Ugh. Now you're talking about 20 year old code...

Originally, `copy_sep` was not preserving sharing, so it is clear that one had to call `copy` on it to preserve sharing.
Now that `copy_sep` preserves sharing, is it still needed ?
At least it is needed for objects, if the row variable is not generic. In that case we allow copying  fields using `copy_sep`, but the row variable, which allow to add extra fields should still be shared with the original.
For variants, it can only be called if the row variable is generic, so that the result of copying it with `copy_sep` or `copy` should be identical.

Reply:
As for `keep`, while it is sound, it does not always make sense to share an extension between incompatible types.
More precisely, `keep` works well for `Subst.typexp`, as it is there used to compute signatures of modules: the original type is no longer used after the substitution, so the compatibility doesn't matter. However, in `copy_sep` where we might want to unify both the original and the copy later on, this is essentially delaying problems, and I prefer to stick to a well-understood theory, which requires correct scoping from the beginning.